Fixing PLD for K2

nksaunders opened the pull request.

nksaunders committed:
fixing for K2

barentsen commented on lightkurve/correctors/pldcorrector.py:
```diff
@@ -179,12 +179,28 @@ def create_design_matrix(self, pld_order=3, pca_components=16, pld_aperture_mask
         # We set the width of all coefficient priors to median flux divided
         # by 3 to prevent the fit from going crazy.
-        prior_sigma = np.nanmedian(self.lc.flux.value)
+        prior_sigma = np.nanstd(self.lc.flux.value) * 10
```
Can you expand on why you think `nanstd` is better than `nanmedian`?  My thinking was that the components of the designmatrix all tend to be normalized (because `pca()` normalizes them), so my reasoning was that we want to allow coefficients to range "roughly between zero and the median flux".

Do we think it is more principled to use `nanstd`?  Would this not  prevent a fit if the median flux is very large compared to the scatter?  Perhaps we could allow for that by setting `prior_mu` of the `append_constant` column equal to the median flux?

barentsen commented:
@nksaunders Can you point me to an example target where the flux-normalization of the regressors appears to improve the detrending?  The step adds documentation and debugging complexity to the PLD corrector, so I'd like to make sure that we can convince ourselves that it is worth having.

nksaunders commented:
Here is my justification for the changes I made here.

## Normalization

The normalization step is separated into `on_target_pix` and `off_target_pix` for three important reasons:
- For K2, we want to normalize out the time-varying signal that is not due to motion noise, as motion noise is the dominant source we need to remove
- For both K2 and TESS, we want to normalize out the time-varying *stellar* signal from on-target pixels to ensure the PLD vectors don't overfit and remove stellar signal
- When we normalize out the stellar signal, we don't want to divide an off-target pixel by an on-target pixel to avoid injecting stellar signal into the background

I typically use K2-199 as a benchmark for these changes, as it is dramatically affected by K2 noise and has easily visible stellar variability. Here I show three cases for K2-199.

### 1. No flux normalization

![download-2](https://user-images.githubusercontent.com/17130840/93922384-89d1ef00-fcad-11ea-9c3b-30844997cdcd.png)

While this initially looks like a good correction qualitatively, and the CDPP is very low, it's important to notice in the middle panel that the pixel series follows the same trend as the spline, fitting out the stellar rotation signal. We don't want that. This is a case of **overfitting** by PLD.

### 2. Only on-target normalization

This is what we want to do for TESS, and trying it on your favorite TESS target will demonstrate that it works well (for instance, WR40 maintains a clear pulsation signal; see section 4 for more details). This will prevent PLD from fitting out stellar signal by normalizing it out of on-target regressors for K2 targets, and should have no effect on TESS by default as there is no overlap between the `aperture_mask` and `pld_aperture_mask` (though it would serve the same purpose as for K2 were someone to pass a custom `pld_aperture_mask`). 

![download-1](https://user-images.githubusercontent.com/17130840/93922950-63608380-fcae-11ea-8eb6-64463ad45675.png)

This has resolved the primary issue present in case 1, and the pixel series no longer fits out the stellar rotation. However, the scatter of the corrected light curve is much more dramatic and the CDPP has spiked. I believe this is because normalizing the on-target pixel regressors (which should be the dominant regressors in the model for K2) down-weights them relative to the background pixel regressors. This remains consistent with EVEREST, which flux-normalizes every pixel (though we've gone back and forth about whether background pixels should be normalized by a light curve that includes on-target pixels, which I believe they should not). This takes us to the third case.

### 3. On-target and off-target normalization

Here, I show the result of separately normalizing on-target and off-target pixels.

![download](https://user-images.githubusercontent.com/17130840/93923349-fbf70380-fcae-11ea-8730-10c5d3cd7c6d.png)

The CDPP is greatly improved by normalizing the background, and the pixel series still does not fit out the stellar signal. I believe this is the ideal case for K2.

### 4. Justification for not normalizing off-target pixels for TESS

The dominant source of noise in TESS FFI cutouts is the scattered light background. This high-amplitude signal should **not** be normalized out from each cadence, as it is a high-amplitude, time-varying, additive signal that should be captured by the pixel design matrix in the absence of a background design matrix. 

Here, I show two summary plots for WR40, a star with distinct stellar variability. First, the case where we treat TESS and K2 identically in the creation of the design matrix, and flux-normalize the background pixel regressors.

![download-4](https://user-images.githubusercontent.com/17130840/93923977-d1597a80-fcaf-11ea-9b95-b05a67850213.png)

This is a terrible correction for two reasons:
- The pulsation signal of WR40 has been completely fit out by PLD
- The background scattered light signal is still present in the corrected light curve

Let's try treating this separately from K2, and not normalize the off-target pixel regressors.

![download-5](https://user-images.githubusercontent.com/17130840/93924428-6ceaeb00-fcb0-11ea-8b81-a0b21425a4ba.png)

Much better! The pixel series has isolated the background scattered light, and the pulsations of WR40 remain in the final corrected light curve.

## Priors

For the issue of priors, I found that with correct normalization, this value is not as important as long as it is finite and reasonable. That being said, I made this change after realizing that in the case of dramatic overfitting (i.e. case 1), setting prior sigma to something lower, and dependent on the standard deviation, helped prevent the pixel components from completely fitting out the stellar signal. There will be cases in which the default aperture masks are not ideal, and I believe this will help us prevent worse overfitting in those cases.

I'll demonstrate this with an example. Here, I show the correction in case 1 (with a prior of 10σ) versus the correction with a looser prior (median flux).

![download-7](https://user-images.githubusercontent.com/17130840/93926018-b0deef80-fcb2-11ea-827b-e1cda0ec2903.png)

The tighter prior appears to have prevented worse overfitting in this case of poor design matrix construction. While this is not terribly demonstrative of the issues we will run into regularly, I think it will be helpful in cases where PLD would otherwise overfit. 

As a final example, I'll show the difference in performance when applied to a correctly normalized design matrix.

![download-9](https://user-images.githubusercontent.com/17130840/93927862-2ea3fa80-fcb5-11ea-886d-1887e042de66.png)

While both of these look like good corrections, and the CDPP is almost identical, the "hockey stick" feature is more pronounced in the correction with the looser prior. This is a very common issue for K2 corrections, and it appears that the tighter prior helps prevent it from blowing up at the beginning and end of the light curve.

I also believe that when normalizing the regressors, the standard deviation makes more sense as a prescription for prior sigma. In my tests, it seems that using 10σ instead of median has no dramatic impact on how well the correction performs, but can prevent overfitting in specific problem cases.

---

Please let me know if you have any questions about any of this.

nksaunders committed:
new mask kwargs; add  mask option

nksaunders commented:
New changes:
- added an `'empty'` option to `tpf._parse_aperture_mask()` (to automatically return the inverse of `'all'`) which I use to create the TESS PLD aperture mask
- set defaults for `pld_aperture_mask`: 
  - TESS: `'empty'`
  - K2: `'threshold'`
- added back `background_aperture_mask` with default value `'background'`, and added it to `.diagnose_masks()`
- added `normalize_background_pixels` kwarg and set its defaults:
  - TESS: `False`
  - K2: `True`
- set the `ylim` argument in `.diagnose()` to be the minimum of either the uncorrected or corrected light curve to ensure neither is cut off in the plot (and made the colors more distinct)

I don't see any case in which the pixels in the `pld_aperture_mask` should not be normalized, so I normalize them in all cases. Let me know if you can think of a reason to include a `normalize_pld_pixels` kwarg.

nksaunders committed:
clean up comments

barentsen commented:
@nksaunders I think I'm happy to merge this!

If you have spare 20 minutes: would it be worth treating pld and background pixels as two separate design matrices?  I ask because I wonder if it'd be helpful to have them show up separately in the `diagnose()` plot.

nksaunders commented:
> would it be worth treating pld and background pixels as two separate design matrices?

I gave this a try, and it seems to work well! Two important considerations:
1. Higher order PLD now works differently. Before, it calculated products of pixels on and off target. Now, it only creates higher order combinations of pixels in the aperture and the background is only first order. That said, it actually seems to have improved the performance.
2. `pca_components` is now more complicated. Should it be flexible to being passed a tuple and parse `(background, pld order 1, pld order 2, ...)` or should it simply be an int that is applied in all cases? Or should we add an additional `background_pca_components` term? I think for simplicity, enforcing that `pca_components` be an int makes the most sense. If a user is delving into the minutia to the point that they want different numbers of PCA components for each order of PLD and the background, they are likely better suited to use the `RegressionCorrector` for its flexibility. Let me know if you have opinions for how it should work.

Other than that, it just required a bit of restructuring of the `create_design_matrix` function, especially in light of the fact that our `pld_aperture_mask` for TESS is empty. Take a look and let me know if anything is unclear.

nksaunders committed:
add back background design matrix (full circle)

nksaunders committed:
fix broken test

nksaunders committed:
bump test

barentsen commented:
I think I'm in favor of requiring `pca_components` to be a single int, which is applied to every DesignMatrix except the spline matrix. It's important to keep it all simple enough to be explainable.

@nksaunders One more idea: is there merit in treating every order of PLD as a separate DesignMatrix? Would it make diagnose more informative? Would it make the application of a single pca_components value more sensible?

nksaunders commented:
@barentsen I tried it, and I'm not sure that it's that useful to separate out the PLD orders into their own design matrices. The middle panel starts to get pretty crowded, and the higher orders are generally just a lower amplitude version of the first order model. If a user wants to inspect a single order, it's pretty straightforward to index it from the `pld.dmc` and `pld.coefficients`.

My vote is to keep a single background matrix and a single PLD matrix for clarity and convenience.

barentsen commented:
Sounds good :+1:
